# Домашнее задание 3
### Линейная и логистическая регрессия

Оформление дз:
Присылайте выполненное на почту `mlphystech@gmail.com` файлы .pdf и .ipynb или ссылку на свой гитхаб.

**Дедлайн: 11 апреля 23:59**

*Зуева Надежда*

ФИВТ МФТИ

Это задание будет посвящено линейной и логистической регрессии.
Мы надеемся, что вы уже знаете, что такое `kaggle`, так как сегодня нам предстоит поработать с самым *классическим* датасетом машинного обучения, а именно $-$ с Титаником!


Цель соревнования $-$ предсказать, кто из пассажиров выживет, а кто, увы, нет.

https://www.kaggle.com/c/titanic

## Теоретические вопросы
### 2 балла (по 0.5 за каждый верный ответ)
1. Что такое линейные модели классификации и регрессии? В чем их основные свойства? Где хорошо работают?
** Мы их используем тогда, когда полагаем, что зависимость между классом объекта(здесь, условно, выживший/нет) и его параметрами линейна, т.е. 
может быть выражена как зависимость вида y=w1x1+w2x2+... Т.е. мы успешно классифицируем объект, если знаем эту зависимость.
Соотвественно, хорошо работают, если зависимость линейна(правда что ли? :) )**
**Постараюсь описать максимально просто:
    У нас есть алгоритм. Как это работает. Мы полагаем, что зависимость Y от признака(ов) линейна – т.е. мы как бы можем построить прямую зависимости. Как мы это делаем? Предположим пока, что у нас есть только 1 признак. Тогда y=w1*x+const. И пусть у нас есть несколько точек. Искомая прямая должна проходить на плоскости ( а при одном признаке у нас пока плоскость ) таким образом, чтобы MSE было минимальным. MSE- это как бы сумма всех (yi-f(xi))^2. Т.е. сумма квадратов разности истинных значений y для каждой точки и тех значений, что по данным признакам вычисляет функция. Теперь как мы вычислим ф-ю?  А мы возьмём и найдём минимум ф-и E(yi-f(xi))^2. Тогда мы здесь получим в итоге прямую искомую, т.е. найдём коэффициент при x, т.е. w1, т.е. вес признака. И тогда мы получим искомую прямую y=f(x). То же самое происходит в многомерном пространстве. Пусть у нас есть n признаков. Тогда в n-мерном пространстве у нас будет задана некоторая прямая с некими коэффициентами при соответствующих осях. Где оси – это признаки. Тогда эти коэффициенты будут представлять собой вектор w=(w1,w2,w3…..wn). Они у нас изначально неизвестны. Но у нас есть точечки(векторы) в пространстве, и нам нужно построить такую прямую, чтобы MSE было минимальным. Как мы действуем.
Должно быть аналогичное плоскому уравнение (yi-f(xi))^2, но только в многомерном пространстве. Берём здесь, по сути, то же самое. Если у нас F(x1,x2,x3,x4..)=w1*x1+w2*x2+… - как раз таки наша функция. Берём и делаем: 
Функция  суммы E(yi-F(xi))0 ( стремится к нулю ), где xi- вектор признаков. Мы дифференцируем её и находим минимум.(Это делается как-то с перемножением матриц, но я пока не до конца осознал, как это там происходит. Но что происходит, осознал.) Таким образом, находятся минимальные веса w1,w2,w3….  Т.е. наиболее удачная зависимость y от тех или иных признаков.**
2. Что такое регуляризация и как она используется? 
**  Бывает у нас такое, что какие-то признаки имеют слишком большие веса, за счёт это моделька получается довольно таки
кривой и неправдоподобной. Что нужно сделать? Не пропустить слишком большие веса!!! Это и есть регуляризация - как бы 
сглаживание значений весов. Как мы это делаем? Вот у нас там была формула, что должно быть минимальным E(yi-F(xi)). А теперь
берём и к этой формуле дописываем + E(wi). И делаем всё то же самое, что и раньше, но находим уже такие веса, при которых
достигается минимум функции E(yi-F(xi))+E(wi) ( E-сумма).Тогда получается, что и изначальная ф-я стремится быть минимальной,
т.е. находится искомая прямая, но при этом и сумма весов, то есть все веса, тоже стремятся быть минимальными.**
3. Что такое MSE? Приведите пример.
**Это сумма квадратов разностей предсказанного y и истинного y для всех точек. 
Например. y=[0,1,4], y_pred=[1,1,1], тогда MSE=(1-0)^2+(1-1)^2+(4-1)^2= 1+0+9=10**
4. Что такое градиентный спуск? Опишите базовый алгоритм работы.
**По сути, это нахождение какого-то минимума ф-и. Как мы поступаем: у нас есть есть градиент - это как бы производная ф-и, но в многомерном простарнстве.
    Грубо говоря, - мы знаем угол наклона ф-и в каждой точечке. Мы начинаем "вышагивание" - берём и перемещаемся под углом всё ниже
    и ниже. ( Мы знаем угол наклона, значит, мы понимаем, куда шагать. ) Шаги здесь представляют из себя как бы перемещение из начальной точечки
    в следующую на какое-то заданное расстояние. Посколько мы всё время спускаемся куда-то вниз, со временем мы будет тупо кружить
    вокруг какого-то минимума ф-и. ( Чем меньше длина "шага", тем больше точность, но больше требуется времени.)**


P.S. просьба не копипастить из Википедии. Это все легко вычисляется и будет караться :)

## Задание 1 (4 балла)
Скачайте файлы `train.csv` и `test.csv`.

Выведите первые 10 строк из `train.csv`

In [16]:
import pandas
from copy import deepcopy
import numpy as np
import pandas as pd

data = pd.read_csv('train.csv') #считайте данные из файла train.csv


In [2]:
data.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Как мы видим, не все данные *

полны*. Например, это видно в столбце `Age`. 
В качестве простейшего решения мы предлагаем заменить пропущенные значения на медианные (функция `numpy.median`)
Вы можете воспользоваться нашим методом или произвести собственный анализ и заменить значения более умным способом (profit: бонусные баллы).

In [17]:
# Replace all the missing ages in the data with the median age
df = deepcopy(data)
df["Age"] = df["Age"].fillna(df["Age"].median())
df.head(10)
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.361582    0.523008   
std     257.353842    0.486592    0.836071   13.019697    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Теперь займемся **оцифровкой признаков**

Кстати, напомните, а зачем нам это нужно? :)

$*$ ваш ответ: $*$ **Так проще. Легче работать с циферками, чем с какими-то бувенными обозначениями.
    Т.е. их легче анализировать, так, например, ф-я describe проводит анализ только для признаков,
    что заданы численно. Это понятно.**
    

Какой параметр здесь бинарный? Оцифруйте его при помощи нулей и единиц.

In [18]:
df.loc[df["Sex"] == "male", "Sex"] = 0
df.loc[df["Sex"] == "female", "Sex"] = 1

Выведите описание нашего датасета после изменений с помощью функции `describe`.
Сколько процентов пассажиров выжило? Каков был средний возраст всех пассажиров?
** Выжило 38.3838...%. Средний возраст 29.36  **

In [5]:
print (df.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  891.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.361582    0.523008   
std     257.353842    0.486592    0.836071   13.019697    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   22.000000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   35.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


Продолжим заполнять пропуски. Рассмотрим графу пунктов отправления `Embarked` 

Там есть типы данных:

C = Cherbourg, Q = Queenstown, S = Southampton, а также `nan`

In [5]:
print (df["Embarked"].unique())

['S' 'C' 'Q' nan]


In [6]:
import collections
print(collections.Counter(df['Embarked']))

Counter({'S': 644, 'C': 168, 'Q': 77, nan: 2})


С учетом последней ячейки заполните пропущенные значения по своему усмотрению. Объясните свой выбор, затем оцифруйте признак `Embarked` 

Например, заменив все на {0,1,2} или {12,42,10}, как хотите :)
Объяснение выбора: ** Есть три варианта, на что менять - на Q, или S, или C. Если мы заменим nan на какую-то
    из этих букв, то она окажет как бы влияние на статистику объектов, имеющих такой же признак. Причём чем
    меньше объектов с этим признаком было изначально, тем большее влияние окажет добавление к ним новых объектов.
    Поэтому, чтобы снизить влияние добавления этих как бы случайных величин ( а его необходим снизить, так как
    мы не знаем настоящий признак добавляемых объектов и они, на самом-то деле, вносят ненужные помехи в статистику)
    надо добавить их к наибольшей группе объектов, здесь к объектам, обладающим признаком S -  644 объекта.**

In [8]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris   0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1      0   
2                             Heikkinen, Miss. Laina   1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1      0   
4                           Allen, Mr. William Henry   0  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [19]:
df["Embarked"] = df["Embarked"].fillna("S")
df.loc[df['Embarked']=='S','Embarked']=1
df.loc[df['Embarked']=='C','Embarked'] =2
df.loc[df['Embarked']=='Q','Embarked'] =3


Разобьем выборку на `X` и `y`, где для`y` будем строить предсказания

In [8]:
print(df.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [20]:
y = pd.Series(df['Survived'])

In [21]:
X = pd.DataFrame(df[['PassengerId','Pclass','Name' ,'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']])# Все эти параметры важны.
#X = pd.DataFrame(df[["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]])

# Здесь id не играет роли - это просто номера человека в таблице. Имя, очевидно, тоже. Каюта -то, может,
# и играет какую-то роль, но её сложно учесть, так что отбросим.

Каким образом можно отобрать фичи, которые будут влиять на наш скор? Тут нам помогут ваш жизненный опыт и ` Recursive Feature Elimination` из библиотеки `Sklearn` 

**Я бы по-хорошему вручную удалил к чёртикам и Ticket, и Cabin, и PassengerId, и Name. Но тут от меня, по-видимому, требуется сначала
эти признаки оцифровать, а потом с помощью RFE сделать вывод о том, что они бесполезны. Так что сначала оцифруем.**

In [31]:
letters=df['Ticket'].unique()
for i in range(len(letters)):
    df.loc[df['Ticket']==letters[i], 'Ticket']=i  # Да, это что-то странное и неэффективное.
letters=df['Cabin'].unique()
for i in range(len(letters)):
    df.loc[df['Cabin']==letters[i], 'Cabin']=i
letters=df['Name'].unique()
for i in range(len(letters)):
    df.loc[df['Name']==letters[i], 'Name']=i
# Поскольку в 'Cabin' очень много NaN-ов - заменив их по-тупому на медианные значения
df["Cabin"] = df["Cabin"].fillna(df["Cabin"].median())

In [33]:
# Так как все эти извращения завершены, поработаем с RFE.

In [55]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression, LinearRegression
model = LinearRegression()  #(#тут можно подбирать параметры для повышения скора)
my_rfe=RFE(estimator=model)
my_rfe.fit(X,y)
print(my_rfe.ranking_)
print(my_rfe.support_)
print(X.columns)
#тут

[3 1 4 1 2 1 1 7 6 5 1]
[False  True False  True False  True  True False False False  True]
Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


**Получились несколько противоречивые результаты. Грубо говоря, он принял за крайне важные все нужные параметры, кроме
возраста и платы. Но возрасту он дал второе место. Полагаю, если бы делал по-умному, а не через медианное значение, получилось
бы лучше. Но Fare он стабильно причисляет к незначащему признаку.**

Проверьте, похожи ли отобранные вами фичи на те, что представлены ниже?

** К сожалению, Fare он не отбирает, хотя должен бы, окаянный. **

In [22]:
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"]

In [23]:
X=df[features]

In [25]:
from sklearn.cross_validation import KFold
kfold = KFold(df.shape[0], n_folds=3, random_state=1)

Выполним предсказания:

In [111]:
predicts = []
for train, test in kfold:
    X_train = X.loc[train]
    y_train=y.loc[train]
    X_test=X.loc[test]
    model=LinearRegression()
    model.fit(X_train, y_train)
    y_pred= model.predict(X_test)
    predicts.append(y_pred)


predicts = np.concatenate(predicts)

#print (predicts)

predicts[predicts > 0.5],predicts[predicts < 0.5] = 1,0

accuracy = np.count_nonzero(df["Survived"] == predicts)/df["Survived"].count()
print ('Скор для линейной регрессии:',accuracy)


Скор для линейной регрессии: 0.783389450056


**Итог: линейная регрессия дала достаточно точные пронозы( не так уж плохо - 0.78)**

## Задание 2 (2 балла)
Теперь попробуем проделать то же самое для **логистической регрессии**

http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

In [26]:
model = LogisticRegression()

Вычислим оценку точности для всех кросс-валидационных фолдов: возвращаем массив баллов из 3-х фолдов

In [169]:
y = pd.Series(df['Survived'])
features = ["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare", "Embarked"] # Мы уже значем, какие параметры наормальные.
X=df[features]
kfold = KFold(df.shape[0], n_folds=3, random_state=1)         #n_folds можно поменять
predicts = []
for train, test in kfold:
    X_train = X.loc[train]
    y_train=y.loc[train]
    X_test=X.loc[test]
    model.fit(X_train, y_train)
    y_pred= model.predict(X_test)
    predicts.append(y_pred)


predicts = np.concatenate(predicts)
predicts[predicts > 0.5],predicts[predicts < 0.5] = 1,0
accuracy = np.count_nonzero(df["Survived"] == predicts)/df["Survived"].count()
scores=pd.Series([predicts==y])[0]

print("Скор для логистической регрессии: " + str(scores.mean()))

Скор для логистической регрессии: 0.792368125701


###### Как бы вы объяснили прирост качества в логистической регрессии? 
Сможете побить наш бейзлайн? :)

## Задание 3 (1.5 балла)
Отправьте ваше решение в тестирующую систему и посмотрите, какое место вы займете

рассмотрим `test.csv`

In [58]:
df_test = pandas.read_csv("test.csv")
print(df_test.describe())

       PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200


Надо вернуть оцифрованные данные к первоначальному виду:

In [40]:
df_test.head()[features]

Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0       3    male  34.5      0      0   7.8292        Q
1       3  female  47.0      1      0   7.0000        S
2       2    male  62.0      0      0   9.6875        Q
3       3    male  27.0      0      0   8.6625        S
4       3  female  22.0      1      1  12.2875        S

In [32]:
#тут ваш код по возвращению оцифрованных параметров в первоначальный вид

# Они уже в первоначальном виде. Оицфруем-ка мы их, пожалуй!


In [59]:
df_test.loc[df_test["Sex"] == "male", "Sex"] = 0
df_test.loc[df_test["Sex"] == "female", "Sex"] = 1
df_test["Embarked"] = df_test["Embarked"].fillna("S")
df_test.loc[df_test['Embarked']=='S','Embarked']=1
df_test.loc[df_test['Embarked']=='C','Embarked'] =2
df_test.loc[df_test['Embarked']=='Q','Embarked'] =3

df_test["Fare"] = df_test["Fare"].fillna(df_test["Fare"].median())
df_test["Age"] = df_test["Age"].fillna(df_test["Age"].median())



In [48]:
df_test.head()[features]  # оцифровали!!! 

Pclass Sex   Age  SibSp  Parch     Fare Embarked
0       3   0  34.5      0      0   7.8292        3
1       3   1  47.0      1      0   7.0000        1
2       2   0  62.0      0      0   9.6875        3
3       3   0  27.0      0      0   8.6625        1
4       3   1  22.0      1      1  12.2875        1

In [68]:
model.fit(df[features], df['Survived'])
predicts = model.predict(df_test[features])

submission = pandas.DataFrame({
    "PassengerId": df_test["PassengerId"],
    "Survived": predicts
})

submission.to_csv("submission.csv", index=False) #вот это отправляется в каггл

Можете попробовать и дальше "поиграться" с параметрами, чтобы подняться в лидерборде. Например, попробовать подобрать параметры в регрессорах. 
Удачи!
    Напоминаем, что справку по любой функции (в том числе и по регрессорам) можно получить, если с наведенным курсором на функцию нажать `Tab`+`Shift`

**Тут, наверное, стоит написать какой-то вывод? Напишем. Линейная классификация - это круто. Как мне кажется, это уже что-то
похожее на настоящее обучение( вообще, кажется, им и является) - мы уже тренируем моделько по-нормальному ( а не понарошку, как с kNN)
Мы действительно устанавливаем прямые зависимости между какими-то признаками и свойствами объекта( жаль только, что 
в реальности эти зависимости не всегда прямые ). Очевидная проблема - мы предполагаем, что зависимости линейны. Прикол в том,
что вряд ли это на самом деле так. В любом случае жизнь - штука сложная, и вряд ли всё-всё можно учесть в линейных коээфициентах.
За счёт этого получается не самая высокая точность. Тем не менее, в случае, когда какие-то признаки связаны линейно - я уверен, это 
идеальный вариант.**